Implement opinion versions in frontend

quevon24 opened the pull request:
## Fixes
<!-- What bugs does this fix? Use this syntax to auto-close the issue: -->
<!-- https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword -->
<!-- E.g.: "Fixes: #XYZ" -->
None

## Summary
<!-- What does this fix, how did you fix it, what approach did you take, what gotchas are there in your code or compromises did you make? -->
This PR introduces a new feature that allows users to view different versions of opinions. 

- Users can now access and view historical versions of opinions. 
- A cached_property named opinions_versions and opinion_main_version has been implemented in the model. This efficiently fetches and caches sub-opinion versions, ordered by creation date, to optimize performance. Also fetches main version of a cluster.
- Up to a maximum of 10 versions are displayed.

## Deployment

**This PR should:**
<!-- The following labels control the deployment of this PR if they’re applied. -->
<!-- Please put an "X" in the box on ones that apply. -->

<!-- Check here if the entire deployment can be skipped -->
<!-- This might be the case for a small fix, a tweak to documentation or something like that. -->
- [ ] `skip-deploy`
<!-- Check here if the web tier can be skipped -->
<!-- This is the case if you're working on code that doesn't affect the front end, like management commands, tasks, or documentation. -->
- [ ] `skip-web-deploy`
<!-- Check here if the deployment to celery can be skipped -->
<!--This is the case if you make no changes to tasks.py or the code that tasks rely on. -->
- [x] `skip-celery-deploy`
<!-- check this if deployment to cron jobs can be skipped -->
<!-- This is the case if no changes are made that affect cronjobs. -->
- [x] `skip-cronjob-deploy`
<!-- Deployment of daemons can be skipped -->
<!-- This is the case if you haven't updated daemons or the code they depend on. -->
- [x] `skip-daemon-deploy`

<!-- **If deployment is required:** -->
<!-- What extra steps are needed to deploy this beyond the standard deploy? -->
<!-- Do scripts need to be run or things like that? -->
<!-- If this is more than a quick thing, a new issue should be created in our infra repo: https://github.com/freelawproject/infrastructure/issues/new (if you don’t have access to it, just put the steps here) -->
<!-- Please use an ordered list or delete this if no special steps are required: -->


<!-- DELETE this section if your PR doesn't require screenshots. -->
<!-- If this changes the front end, please include desktop and mobile screenshots or videos showing the new feature. -->
<details closed>
<summary><h2>Screenshots</h2></summary>
<details open>
<summary><h4>Desktop</h4></summary>
<img width="1387" height="734" alt="image" src="https://github.com/user-attachments/assets/1bac3fec-f112-4ec5-bd1f-2c478c70bc11" />
In this screenshot you see the same dates because i cloned the data and the date_created value is almost the same for each version



</details>
<details open>
<summary><h4>Mobile</h4></summary>



</details>
<!-- END DELETE -->

<!-- Thank you for contributing and filling out this form! -->

mlissner commented:
This is looking pretty good. One thing I don't see, that I think is important, is some explanation of why we have so many versions. In an earlier version of this there was a page that listed the older versions. Maybe something like that would have space to explain the situation? Another option could be a modal that the button opens which shows the versions, but that might not scale well if there's more than one or two (the current design has this problem too).

mlissner commented:
I like the direction, but I wouldn't click "Download PDF" if I was trying to get to an older version, and I wouldn't expect a download button to take me to an HTML page for an earlier version (I'd expect a PDF). Hm.

quevon24 committed:
feat(opinins): add property to get opinions versions

quevon24 committed:
feat(opinions): improve readability in get_downloads_context
Display opinions versions in download button

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

flooie committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

flooie committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

flooie committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

flooie committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
fix(opinion_page): add rel="noopener noreferrer" to external links

- vLex sponsor link
- SCDB case detail link

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge remote-tracking branch 'origin/main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
fix(opinions): enable PDF tab when opinion has local_path but no download_url

quevon24 committed:
feat(opinions): show PDF accordion for multiple opinions, improve version labels

quevon24 commented:
I implemented a collapsible accordion in the PDF tab when a cluster has multiple independent opinions with PDFs (e.g. lead, dissent, concurrence). Clusters with a single PDF or where all opinions are versions of the same document continue to display the single PDF view as before. Also updated the version labels in the Download PDF dropdown to show the opinion type and date.

Dropdown:
<img width="1246" height="1050" alt="image" src="https://github.com/user-attachments/assets/72dfca79-f672-4b00-a0ee-4df339de362c" />

Collapsible:
<img width="1246" height="1050" alt="image" src="https://github.com/user-attachments/assets/1f88440f-5bd2-4199-924e-570a5e6702e8" />

quevon24 committed:
fix(tests): update pdf_path key to download_file_path

quevon24 committed:
fix(opinions): align download dropdown to right to prevent mobile overflow

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

grossir commented on cl/opinion_page/templates/includes/add_download_button.html:
```diff
@@ -41,5 +41,14 @@
         {% endfor %}
+    {% if cluster.opinions_versions %}
+        <li role="separator" class="divider"></li>
+        <li><a href="#"><b>Versions</b></a></li>
```
This could be a `  <li class="dropdown-header">Versions</li>`  already packaged on bootstrap

grossir commented on cl/opinion_page/views.py:
```diff
@@ -898,29 +898,52 @@ def get_attachment_values(
+        download_file_path = cluster.filepath_pdf_harvard.url
+        embeddable_pdfs.append(
+            {
+                "label": "Harvard PDF",
```
I think the label should be something like "Case Law Access Project scan" or something, to be consistent with other usages

<img width="886" height="719" alt="Image" src="https://github.com/user-attachments/assets/41d9e026-935f-4e5b-881a-6a9f2848c6e1" />

grossir commented on cl/search/models.py:
```diff
@@ -2813,6 +2813,18 @@ def ordered_opinions(self):
         # If there's only one or no sub-opinions, return the main opinion
         return sub_opinions
 
+    @cached_property
+    def opinion_main_version(self):
+        # Fetch opinion main version of the cluster
+        return self.sub_opinions.filter(main_version__isnull=True)
```
We should type annotate the return or change the name or add a better docstring, since the name makes it look like it will return a singular opinion, when a cluster with opinion types may return many.

grossir reviewed (commented):
Some small changes, it's looking good

quevon24 committed:
Merge branch 'main' into 5451-show-opinion-versions-in-opinion-page

quevon24 committed:
refactor(opinion_page): address review feedback on opinion versions

- use bootstrap dropdown-header for the Versions label
- rename Case Law scan labels to Case Law Access Project Scan
- annotate and document opinion_main_version return type

grossir reviewed (approved):
LGTM
